### ISSUE TITLE
WMS performance issue with non-square pixels

### ISSUE THREAD

**Issue by foggel** [mapserver/mapserver#5625]
Scale-calculation is based on the x-axis-cellsize. It is possible to request maps where the x-cellsize and the y-cellsize are totally different. 

If a WMS-request has the parameters: BBOX=1,1,2,1000&WIDTH=100&HEIGHT=100 then there is a big difference in cellsize. 

In this case the scaledenominator will be low (28.4) because maxx-minx distance is only 1 unit and layers supposed to be rendered at zoomed-in scales will be enabled.  The distance on the y-axis can be extreme, in the example above 999 units, making the area of the bounding box sent to the underlying datasource large.

Many WMS-services can easily be overloaded by having extreme distances on the y-axis and small distances on the x-axis. If the WMS-request is using a SRS/CRS different from the datasources native SRS the envelope send to the datasource can be extremely large.

Question: Would it be possible to change the scale calculation to either
1) base the scale on the largest cellsize
2) base the scale on the pixels diagonal cellsize

Example: using the largest cellsize in scale calculation

```
--- mapscale.c.orig     2018-07-05 10:38:46.102919675 +0200
+++ mapscale.c.new      2018-07-05 10:49:59.276283944 +0200
@@ -76,7 +76,8 @@
 */
 int msCalculateScale(rectObj extent, int units, int width, int height, double resolution, double *scale)
 {
-  double md, gd, center_y;
+  double md, gd, center_y, center_x;
+  double cellsize_x, cellsize_y;

   /* if((extent.maxx == extent.minx) || (extent.maxy == extent.miny))   */
   if(!MS_VALID_EXTENT(extent)) {
@@ -97,11 +98,20 @@
     case(MS_NAUTICALMILES):
     case(MS_INCHES):
     case(MS_FEET):
-      center_y = (extent.miny+extent.maxy)/2.0;
-      md = (width-1)/(resolution*msInchesPerUnit(units, center_y)); /* remember, we use a pixel-center to pixel-center extent, hence the width-1 */
-      gd = extent.maxx - extent.minx;
+      cellsize_x = MS_CELLSIZE(extent.minx, extent.maxx, width);
+      cellsize_y = MS_CELLSIZE(extent.miny, extent.maxy, height);
+      if(cellsize_x >= cellsize_y) {
+        center_y = (extent.miny+extent.maxy)/2.0;
+        md = (width-1)/(resolution*msInchesPerUnit(units, center_y)); /* remember, we use a pixel-center to pixel-center extent, hence the width-1 */
+        gd = extent.maxx - extent.minx;
+      }else {
+        center_x = (extent.minx+extent.maxx)/2.0;
+        md = (height-1)/(resolution*msInchesPerUnit(units, center_x));
+        gd = extent.maxy - extent.miny;
+      }
       *scale = gd/md;
       break;
+
     default:
       *scale = -1; /* this is not an error */
       break;

```

Maybe this issue is related: https://trac.osgeo.org/mapserver/ticket/3676

**Comment by sdlime** [comment]
I wonder what sort of negative impacts this will have on existing configurations - I'd bet not much. Question is what method makes the most sense. Have you tried the diagonal approach? --Steve

**Comment by foggel** [comment]
Thanks for responding.  My goal is to protect my servers from bad requests/requests with extreme pixel aspect ratios, and this is currently an issue for me. I don't really understand the use cases of non square pixels so I can't say what is the most sensible approach. I was hoping someone else could shed some light.

By using the axis with the largest cellsize only existing configurations where the y-cellsize is larger is affected. I'm guessing the opposite is more common. By using the diagonal distance in calculations all requests with non square pixels will be affected.

By diagonal i meant to replace: 
``` 
gd = extent.maxx - extent.minx 
with
gd = sqrt(0.5*pow((extent.maxx-extent.minx),2) + pow((extent.maxy-extent.miny),2))
```

**Comment by jratike80** [comment]
Non-square pixels are required by the WMS standard. Initial implementation in Mapserver seems to be due to https://trac.osgeo.org/mapserver/ticket/728. Back then the map->cellsize was set to be the average of the x and y cellsize.

**Comment by aperi2007** [comment]
The documentation report

MS_NONSQUARE [yes|no]

has no effect for WMS
